Simpler Python Parameter class

codingS3b opened the pull request:
This PR gets rid of `UiParameter`, `LinearScaledParameter` and `LogScaledParameter` and instead introduces a single `Parameter` class which has the option to pass in arbitrary conversion functions to handle different scales on the UI and on the internal PIC scale of each parameter.
It also makes use of the `pint` package for easier unit conversion.

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -5,17 +5,41 @@
-    Base class for exposable parameters.
+    Parameter that can have a corresponding representation as a jupyter widget
+    within interactive jupyter notebooks.
+    The widget allows the user to adjust values as he likes but this class
```
he/she

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -5,17 +5,41 @@
+    this class. Either as MultiRangeSlider or as MultiSelection or as Checkbox
+    depending on the 'values' (discrete) or 'range' (continuous)
+
+    Stepsizes will be handled by the widget representation, not here.q
```
~~q~~

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
+            Implicitely we assume the pic unit is SI. If this is not possible,
+            the converter needs to translate those scales.
+        default: float, int or string
+            The default value for this parameter (on the UI scale) used when\
```
\ ?

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
+        label: string [optional]
+            Overwrite the name for UI
+        pic_to_SI: callable, e.g. lambda function
+            Specifies how to transform a value on the pic scale to an\
```
\ ?

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
+                self.values = [values]
+            if not self.values:
+                self.values = [self.default]
+                print("WARNING: Values attribute can not be an empty\
```
I think
```python
print("WARNING: Values attribute can not be an empty"
      "iterable! Settting values to: ", self.values)
```

should work

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
+            res = all([v in self.values for v in vals])
+            if not res:
+                raise ValueError(
+                    "Invalid values found! Values should be elements of\
```
```python
"Invalid values found! Values should be elements of "
"self.values!")

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
+            # check for valid range
+            res = all([self.range[0] <= v <= self.range[1] for v in vals])
+            if not res:
+                raise ValueError("Invalid values found! Values should be\
```
see above

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -25,417 +49,155 @@ def __init__(self, name, ptype, unit, default, value=None, dtype=None):
-        # since base class UiParameter now calls the set_value() to carry out
-        # linear scaling of the value, we need to define the scale_factor
-        # before entering the constructor since it is used in set_value()
+        # todo check if vals are in specified range
```
open todo :-)

codingS3b commented:
Thanks for reviewing, I fixed the linebreaks and incorporated the other changes.
Would it be alright to also include the updated params.py for LWFA example in this PR that already uses the new `Parameter` class?

ax3l commented:
Good point, go ahead and keep it in consistent state.

ax3l commented on lib/python/picongpu/input/parameters.py:
```diff
@@ -32,8 +32,7 @@ class Parameter(object):
 
-    Stepsizes will be handled by the widget representation, not here.q
-
+    Stepsizes will be handled by the widget representation, not here.
```
ups, step sizes is two words

ax3l replied:
Probably "stepping" is easier to understand?

codingS3b replied:
Agreed.
